### DISCUSSION TITLE
Type for Request object

### DISCUSSION THREAD

**Discussion by text2n** [FoalTS/foal#1084] [Q&A] [answered]
I have a case where i need to pass the ctx.request object to a function. But I am not able to find an exposed Request type/interface inside foalts. Request is defined in core/http/context.d.ts, but not exported.  As an alternative I tried using Request interface from @types/express. But typescript is throwing incompatibility error. Can someone pls suggest me how to use type for Request object?

**Comment by LoicPoullain** [comment, marked as answer]
Hi @text2n 👋 

Can it work if you define a `Request` type like this in your code?
```typescript
type Request = Context['request']
```

**Comment by text2n** [comment]
@LoicPoullain The issue is when we have a function which accepts request as parameter. For example, the below hypothetical function accepts request object (req) as parameter. if we are using typescript, to define the type of "req" we need access to Request interface.  I am sorry if i am missing something. I am bit new to typescript.

```
function getIp(req: Request) {
   return req.ip;
}

const ip = getIp(Context.request);
```

**Reply by LoicPoullain** [reply]
Yes, the above example gives a way to redefine the `Request` type from the `Context` so that it can be used in your function. In your case it would look like this:

```typescript
import { Context } from '@foal/core';

type Request = Context['request']

function getIp(req: Request) {
   return req.ip;
}

const ip = getIp(ctx.request);
```
Or another option:
```typescript
import { Context } from '@foal/core';

function getIp(req: Context['request']) {
   return req.ip;
}

const ip = getIp(ctx.request);
```

**Reply by text2n** [reply]
@LoicPoullain It works, thank you so much.
